docs/backend-system/architecture/03-services.md

```diff
@@ -1,103 +1,227 @@
+
+The plugin metadata service is the base for all plugin specific customizations for services. For example, the default implementation of the plugin scoped logger service uses the plugin metadata service to attach the plugin ID as a field in all log messages:
+
+```ts
+export const loggerFactory = createServiceFactory({
+  service: coreServices.logger,
+  deps: {
+    rootLogger: coreServices.rootLogger,
+    pluginMetadata: coreServices.pluginMetadata,
+  },
+  factory({ rootLogger, pluginMetadata }) {
+    return rootLogger.child({ plugin: pluginMetadata.getId() });
+  },
+});
+```
+
+## Root Context for Service Factories
+
+Some service may benefit from having a context that is shared across all instances of a service. This of course only applies to plugin scoped services, as root scoped services only ever have a single instance. The root context could for example be used for sharing a common connection pool for database access, generated secrets for development, or any other kind of shared facility. Note that you should not use this to share state between plugins in production, as that would be a violation of the [plugin isolation rule](./04-plugins.md#rules-of-plugins).
+
+The root context is defined as part of the service factory by passing the `createRootContext` option:
+
+```ts
+export const fooFactory = createServiceFactory({
+  service: fooServiceRef,
+  deps: { rootLogger: coreServices.rootLogger, bar: barServiceRef },
+  createRootContext({ rootLogger }) {
+    return new FooRootContext(rootLogger);
+  }
+  factory({ bar }, ctx) {
+    return ctx.forPlugin(bar)
+  },
+});
+```
+
+Whatever value is returned by the `createRootContext` function will shared and passed as the second argument to each invocation of the `factory` function. That way you can create a shared context that is used in the creation of each plugin instance. Unlike the `factory` function, the `createRootContext` function will only receive root scoped services as its dependencies, but just like the `factory` function, it can also be `async`.
+
+## Default Service Factories
+
+There are a lot of services that are installed in any standard Backstage backend instance by default. You can expect on these services to always exist, and do not need to take any additional steps to make them available. This is not necessarily true for services that you import from external packages, as the user of your plugin or module might not have installed a factory for that service in their backend. In order to avoid having to ask integrators of your plugin to install a service factory for a service that you depend on, it is possible to define a default factory for a service.
```

Comment:
```suggestion
There are a lot of services that are installed in any standard Backstage backend instance by default. You can expect these services to always exist, and do not need to take any additional steps to make them available. This is not necessarily true for services that you import from external packages, as the user of your plugin or module might not have installed a factory for that service in their backend. In order to avoid having to ask integrators of your plugin to install a service factory for a service that you depend on, it is possible to define a default factory for a service.
```